Fixes a few bugs with greyscale stacks.

TemporalOroboros opened the pull request:
<!-- Write **BELOW** The Headers and **ABOVE** The comments else it may not be viewable. -->
<!-- You can view Contributing.MD for a detailed description of the pull request process. -->

## About The Pull Request
- Fixes greyscale floor tiles merging regardless of their materials.
- Fixes greyscale floor tiles voiding materials when splitting the stack.
- Fixes greyscale floor tile stacks being created with no mats_per_unit and only enough custom materials for a single unit.
- Fixes greyscale tile flooring being created with the wrong amount of materials.
- Fixes greyscale tile flooring not producing floor tiles/producing floor tiles with 0 units.


<!-- Describe The Pull Request. Please be sure every change is documented or this can delay review and even discourage maintainers from merging your PR! -->

## Why It's Good For The Game
- Fixes #50682, closes #54653, closes #52707
- Fixes #50678 
- Fixes #50679
- Fixes #50680

<!-- Please add a short description of why you think these changes would benefit the game. If you can't justify it in words, it might not be worth adding. -->

## Changelog
:cl:
fix: Fixed various bugs involving generic material floortiles. They're actually usable now.
/:cl:

<!-- Both :cl:'s are required for the changelog to work! You can put your name to the right of the first :cl: if you want to overwrite your GitHub username as author ingame. -->
<!-- You can use multiple of the same prefix (they're only used for the icon ingame) and delete the unneeded ones. Despite some of the tags, changelogs should generally represent how a player might be affected by the changes rather than a summary of the PR's contents. -->

TemporalOroboros committed:
Fixes a few bugs with greyscale stacks.

- Fixes greyscale floor tiles merging regardless of their materials.
- Fixes greyscale floor tiles voiding materials when splitting the stack.
- Fixes greyscale floor tile stacks being created with no mats_per_unit and only enough custom materials for a single unit.
- Fixes greyscale tile flooring being created with the wrong amount of materials.
- Fixes greyscale tile flooring not producing floor tiles/producing floor tiles with 0 units.

Rohesie commented on code/datums/components/storage/concrete/stack.dm:
```diff
@@ -39,7 +39,7 @@
 			S.use(can_insert, TRUE)
 			return TRUE
-	return ..(S.change_stack(null, can_insert), override)
+	return ..(S.split_stack(null, can_insert), override)
```
```suggestion
	I = S.split_stack(null, can_insert)
	return ..()
```
Changed args are automatically passed down correctly.

This is really bad practice, to alter the value that goes down the line, but eh, it was already like this.

Rohesie commented on code/game/objects/items/stacks/stack.dm:
```diff
@@ -58,16 +48,16 @@
-			mats_per_unit[SSmaterials.GetMaterialRef(i)] = in_process_mat_list[i]
-			custom_materials[i] *= amount
+
+	if(mats_per_unit?.len)
+		set_mats_per_unit(mats_per_unit, 1)
+	else if(custom_materials?.len)
```
```suggestion
	if(length(mats_per_unit))
		set_mats_per_unit(mats_per_unit, 1)
	else if(length(custom_materials))
```

Rohesie commented on code/game/objects/items/stacks/stack.dm:
```diff
@@ -84,6 +74,31 @@
+	set_custom_materials(mats_per_unit, amount)
+
+/obj/item/stack/on_grind()
+	for(var/i in 1 to grind_results.len) //This should only call if it's ground, so no need to check if grind_results exists
```
```suggestion
	for(var/i in 1 to length(grind_results)) //This should only call if it's ground, so no need to check if grind_results exists
```

Rohesie commented on code/game/objects/items/stacks/stack.dm:
```diff
@@ -242,11 +257,12 @@
-				for(var/i in custom_materials)
-					used_materials[SSmaterials.GetMaterialRef(i)] = R.req_amount / R.res_amount * (MINERAL_MATERIAL_AMOUNT / custom_materials.len)
-				O.set_custom_materials(used_materials)
+			if(R.applies_mats && mats_per_unit?.len)
```
```suggestion
			if(R.applies_mats && length(mats_per_unit))
```

Rohesie commented on code/game/objects/items/stacks/stack.dm:
```diff
@@ -395,22 +422,22 @@
-	if(istype(AM, merge_type) && !AM.throwing)
-		merge(AM)
+/obj/item/stack/Crossed(atom/movable/crossing)
+	if(can_merge(crossing) && !crossing.throwing)
```
```suggestion
	if(!crossing.throwing && can_merge(crossing))
```
Cheaper check first whenever possible.

Rohesie commented on code/game/turfs/open/floor.dm:
```diff
@@ -324,11 +324,12 @@
 
 /turf/open/floor/material/has_tile()
-	return custom_materials.len
+	return custom_materials?.len
```
```suggestion
	return length(custom_materials)
```

Rohesie commented on code/game/turfs/open/floor.dm:
```diff
@@ -324,11 +324,12 @@
-	for(var/i in custom_materials)
-		var/datum/material/M = i
-		new M.sheet_type(src, FLOOR(custom_materials[M] / MINERAL_MATERIAL_AMOUNT, 1))
+	var/obj/item/stack/tile = ..()
+	tile.set_mats_per_unit(custom_materials, 1)
+	return tile
```
```suggestion
	. = ..()
	if(.)
		var/obj/item/stack/tile = .
		tile.set_mats_per_unit(custom_materials, 1)
```
It could runtime, else.

TemporalOroboros committed:
Apply suggestions from code review

- A bunch of ?.len -> length() changes.
- Removes cursed ..(foo()) syntax.
- Properly uses . in `/turf/open/floor/material/proc/spawn_tile`

Co-authored-by: Rohesie <rohesie@gmail.com>
